src/linalg/basic/matrix.rs

```diff
@@ -182,36 +233,65 @@ impl<'a, T: Debug + Display + Copy + Sized> fmt::Display for DenseMatrixMutView<
 impl<T: Debug + Display + Copy + Sized> DenseMatrix<T> {
     /// Create new instance of `DenseMatrix` without copying data.
     /// `values` should be in column-major order.
-    pub fn new(nrows: usize, ncols: usize, values: Vec<T>, column_major: bool) -> Self {
-        DenseMatrix {
-            ncols,
-            nrows,
-            values,
-            column_major,
+    pub fn new(
+        nrows: usize,
+        ncols: usize,
+        values: Vec<T>,
+        column_major: bool,
+    ) -> Result<Self, Failed> {
+        debug_assert!(
+            nrows * ncols == values.len(),
+            "Instantiatint DenseMatrix requires nrows * ncols == values.len()"
+        );
+        let data_len = values.len();
+        if nrows * ncols != values.len() {
+            Err(Failed::input(&format!(
+                "The specified shape: (cols: {ncols}, rows: {nrows}) does not align with data len: {data_len}"
+            )))
+        } else {
+            Ok(DenseMatrix {
+                ncols,
+                nrows,
+                values,
+                column_major,
+            })
         }
     }
 
     /// New instance of `DenseMatrix` from 2d array.
-    pub fn from_2d_array(values: &[&[T]]) -> Self {
+    pub fn from_2d_array(values: &[&[T]]) -> Result<Self, Failed> {
         DenseMatrix::from_2d_vec(&values.iter().map(|row| Vec::from(*row)).collect())
     }
 
     /// New instance of `DenseMatrix` from 2d vector.
-    pub fn from_2d_vec(values: &Vec<Vec<T>>) -> Self {
-        let nrows = values.len();
-        let ncols = values
-            .first()
-            .unwrap_or_else(|| panic!("Cannot create 2d matrix from an empty vector"))
-            .len();
-        let mut m_values = Vec::with_capacity(nrows * ncols);
-
-        for c in 0..ncols {
-            for r in values.iter().take(nrows) {
-                m_values.push(r[c])
+    pub fn from_2d_vec(values: &Vec<Vec<T>>) -> Result<Self, Failed> {
+        debug_assert!(
+            !(values.is_empty() || values[0].is_empty()),
+            "Instantiating DenseMatrix requires a non-empty 2d_vec"
+        );
+
+        if values.is_empty() || values[0].is_empty() {
+            Err(Failed::input(
+                "The 2d vec provided is empty; cannot instantiate the matrix",
+            ))
+        } else {
+            let nrows = values.len();
```

Comment:
this is inefficient, why would you make a copy of the values into `m_values` when you already have them in memory passed as a reference? Smartcore aims to minimize copying.
All these changes will make instantiation more costly and slower. 

I see also some redundancy, why are doing: 
```
                values
                .first()
                .unwrap_or_else(|| {
                    panic!("Invalid state: Cannot create 2d matrix from an empty vector")
```
when we already checked  `if values.is_empty() || values[0].is_empty()` above?

The `else` should just be:
```
else {
    DenseMatrix::from_2d_vec(&values.iter().map(|row| Vec::from(*row)).collect())
}
```
Please be considerate, adding useless for-loops is never a good choice.

Reply:
Sorry this is wrong. it is like this also in the current version.